### PR TITLE
Add exception to environment variable substitution for :_ and :__ 

### PR DESCRIPTION
**Pull request by wilvk**
Fixes https://github.com/npm/npm/issues/20159

Currently server-specific environment variables are being converted to hyphens with the regex replace, thus causing them to be inaccessible for things like credentials.

By running the command (in Linux):

```
env npm_config_//nexus:8081/nexus/content/repositories/npm-test/:_password=YWRtaW4xMjM= bash
```

An environment variable is present such as:

```
npm_config_//nexus:8081/nexus/content/repositories/npm-test/:_password=YWRtaW4xMjM=
```

this is interpreted as the config variable:

```
{ '//nexus:8081/nexus/content/repositories/npm-test/:-password': 'YWRtaW4xMjM=' }
```

Where the password component is incorrect (`-password` instead or `_password`). 
This prevents the password from being set correctly for use later on in `/lib/config/get-credentials-by-uri.js`.

The same applies for other environment variables such as `__auth`.

The replacement I've used is a bit 'clunky' so if anyone wants to clean it up a bit, feel free.

### CODE REVIEW

**Comment by wilvk** [comment]
bump - any updates/feedback on this PR?

File: lib/config/core.js [hunk 1]
```diff
@@ -353,9 +353,15 @@ Conf.prototype.addEnv = function (env) {
...
       // leave first char untouched, even if
       // it is a '_' - convert all other to '-'
+      // except for if there is :_ or :__ as
+      // these are used for _password and __auth
       var p = k.toLowerCase()
-               .replace(/^npm_config_/, '')
-               .replace(/(?!^)_/g, '-')
+              .replace(/^npm_config_/, '')
+              .replace(/:_/, 'PLACEHOLDERONE')
+              .replace(/:__/, 'PLACEHOLDERTWO')
```
Comment (iarna) [inline lib/config/core.js]: this second regexp will never match because `:__ABC` will be replaced with `PLACEHOLDERONE_ABC`

File: lib/config/core.js [hunk 2]
```diff
@@ -353,9 +353,15 @@ Conf.prototype.addEnv = function (env) {
...
+              .replace(/(?!^)_/g, '-')
```
Comment (iarna) [inline lib/config/core.js]: I wonder if the placeholder business could be ditched in preference for:

`.replace(/(?!^)(⁈[_:])_/g, '-')`

That is, only change `_` to `-` if it's not preceded by the start of the string, a `:` or a `_` . It's not quite the same, granted…

**Review by iarna — CHANGES_REQUESTED** [review]
Ok, so beyond the implementation nits, a change like this would need docs and tests too.

But beyond that I'm concerned that making this kind of change to processing will mess up someone else's weird edge case. As there is a work around for your use case that makes me inclined to play things safe here and not take this.

The work around, and in fact, how we encourage people to configure systems like this is:

Add the following to your `npmrc`:

```
 //nexus:8081/nexus/content/repositories/npm-test/:_password=${NEXUS_PASSWORD}
```

Then set the `NEXUS_PASSWORD` env var, eg

```
env NEXUS_PASWORD=YWRtaW4xMjM= bash
```

**Comment by iarna** [comment]
Oh, and if you'd like to chat more about this in real time I encourage you to drop by #js-npm on the https://package.community/ Discord.

**Comment by wilvk** [comment]
Thanks for the reply @iarna , I wasn't aware that variable substitution worked in-line for `.npmrc`. The `$NEXUS_PASSWORD` approach seems like the best solution so will go with that for my stuff.

Will implement the docs, tests and suggested changes for if/when it is decided to go ahead with this PR.

Cheers,

**Comment by zkat** [comment]
Hi! We're moving repos to https://github.com/npm/cli/pulls! See [our blog post about the migration to npm.community](https://blog.npmjs.org/post/175587538995/announcing-npmcommunity) for details. As such, we're closing all active PRs on this repo.

Could you please re-open this PR against `npm/cli` instead? We're still interested in this patch so we hope you will! We'll continue discussion over there once it's moved. 💚
